pkg/error_code/error_code.go

```diff
@@ -31,42 +36,117 @@
-type RegisteredCode string
-
-const (
-	// InternalCode means the operation placed the system is in an inconsistent or unrecoverable state.
-	// Essentially a handled panic.
-	// This is exactly the same as a HTTP 500, so it is not necessary to send this code over HTTP.
-	InternalCode RegisteredCode = "internal"
-	// InvalidInputCode is a validation failure of an input.
-	// The response will indicate the exact input issue.
-	InvalidInputCode RegisteredCode = "input"
-	// NotFoundCode indicates a resource is missing.
-	// The response can indicate what resource is missing.
-	// However, when a single resource is requested, that may be implicit.
-	NotFoundCode RegisteredCode = "missing"
+// CodeStr is a representation of the type of a particular error.
+// The underlying type is string rather than int.
+// This enhances both extensibility (avoids merge conflicts) and user-friendliness.
+// A CodeStr can have dot separators indicating a hierarchy.
+type CodeStr string
+
+func (str CodeStr) String() string { return string(str) }
+
+// A Code has a CodeStr representation.
+// It is attached to a Parent to find metadata from it.
+// The Meta field is provided for extensibility: e.g. attaching HTTP codes.
+type Code struct {
+	// CodeStr contains the full dot-separted path.
+	// This allows for equality comparison based on CodeStr.
+	CodeStr CodeStr
+	Parent  *Code
+}
+
+// NewCode create a new top-level code.
+// A top-level code must not contain any dot separators: that will panic
+// Most codes should be created from hierachy with the Child method.
+func NewCode(codeRep CodeStr) Code {
+	code := Code{CodeStr: codeRep}
+	if err := code.checkCodePath(); err != nil {
+		panic(err)
+	}
+	return code
+}
+
+// Child creates a new code from a parent.
+// childStr must include the parent codes with dot-separation for documentation purposes.
+// An incorrect parent reference in the string panics.
+func (code Code) Child(childStr CodeStr) Code {
+	child := Code{CodeStr: childStr, Parent: &code}
+	if err := child.checkCodePath(); err != nil {
+		panic(err)
+	}
+	return child
+}
+
+// MetaData is a pattern for attaching meta data to codes and inheriting it from a parent.
+// See MetaDataFromAncestors.
+// This is used to attach a HTTP code to a Code.
+type MetaData map[CodeStr]interface{}
+
+// MetaDataFromAncestors looks for meta data starting at the current code.
+// If not found, it traverses up the hierarchy
+// by looking for the first ancestor with the given metadata key.
+// This is used in the HTTPCode implementation to inherit the HTTP Code from ancestors.
+func (code Code) MetaDataFromAncestors(metaData MetaData) interface{} {
+	if existing, ok := metaData[code.CodeStr]; ok {
+		return existing
+	}
+	if code.Parent == nil {
+		return nil
+	}
+	return (*code.Parent).MetaDataFromAncestors(metaData)
+}
+
+var httpMetaData = make(MetaData)
+
+// SetHTTP adds a HTTP code to the meta data
```

Comment:
The original RFC uses "a". https://tools.ietf.org/html/rfc2616. To my ears, "H" is not a vowel sound.

Reply:
oh, interesting. I'm curious how you pronounce 'H' 😳 Another interesting finding is that RFC2616 does use 'a HTTP', but all RFCs after RFC2616 use 'an HTTP'.

Reply:
a/an has nothing to do with the written letters, and only to do with the spoken sound. The spoken [noun H starts with the phonetic eɪ](https://dictionary.cambridge.org/us/dictionary/english/h-h). This is in fact a vowel sound like "aim". Part of my confusion was around articles incorrectly stating that `an university` is because of the `u` is a long vowel sound, but it is because `u` is a consonant (y) vowel combination.